### PR TITLE
[SPARK-21365][PYTHON] Deduplicate logics parsing DDL type/schema definition

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This PR deals with four points as below:

- Reuse existing DDL parser APIs rather than reimplementing within PySpark

- Support DDL formatted string, `field type, field type`.

- Support case-insensitivity for parsing.

- Support nested data types as below:

  **Before**
  ```
  >>> spark.createDataFrame([[[1]]], "struct<a: struct<b: int>>").show()
  ...
  ValueError: The strcut field string format is: 'field_name:field_type', but got: a: struct<b: int>
  ```

  ```
  >>> spark.createDataFrame([[[1]]], "a: struct<b: int>").show()
  ...
  ValueError: The strcut field string format is: 'field_name:field_type', but got: a: struct<b: int>
  ```

  ```
  >>> spark.createDataFrame([[1]], "a int").show()
  ...
  ValueError: Could not parse datatype: a int
  ```

  **After**
  ```
  >>> spark.createDataFrame([[[1]]], "struct<a: struct<b: int>>").show()
  +---+
  |  a|
  +---+
  |[1]|
  +---+
  ```

  ```
  >>> spark.createDataFrame([[[1]]], "a: struct<b: int>").show()
  +---+
  |  a|
  +---+
  |[1]|
  +---+
  ```

  ```
  >>> spark.createDataFrame([[1]], "a int").show()
  +---+
  |  a|
  +---+
  |  1|
  +---+
  ```

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/api/python/PythonSQLUtils.scala [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.api.python
+
+import org.apache.spark.sql.catalyst.parser.CatalystSqlParser
+import org.apache.spark.sql.types.DataType
+
+private[sql] object PythonSQLUtils {
+  def parseDataType(typeText: String): DataType = CatalystSqlParser.parseDataType(typeText)
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/api/python/PythonSQLUtils.scala]: Without this, I should do something like ...

```
getattr(getattr(sc._jvm.org.apache.spark.sql.catalyst.parser, "CatalystSqlParser$"), "MODULE$").parseDataType("a")
```

File: python/pyspark/sql/tests.py [hunk 2]
```diff
@@ -1246,6 +1246,31 @@ def test_struct_type(self):
         with self.assertRaises(TypeError):
             not_a_field = struct1[9.9]
 
+    def test_parse_datatype_string(self):
+        from pyspark.sql.types import _all_atomic_types, _parse_datatype_string
+        for k, t in _all_atomic_types.items():
+            if t != NullType:
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests.py]: So, if I haven't missed anything, this PR drops the support the type parsing `null`. I guess it is almost seldom that we explicitly set the type with `null`. Also, IIRC, we will support `NullType` via `void` (SPARK-20680) soon as a workaround.

**Comment by HyukjinKwon** [comment]
cc @cloud-fan, @felixcheung and @zero323 who I remember I talked about this few times.

**Comment by HyukjinKwon** [comment]
cc @cloud-fan, @felixcheung and @zero323 who I remember I talked about similar issues few times.

**Comment by felixcheung** [comment]
Add
@gatorsmile 
@holdenk

**Comment by HyukjinKwon** [comment]
retest this please

File: python/pyspark/sql/types.py [hunk 3]
```diff
@@ -806,43 +786,43 @@ def _parse_datatype_string(s):
...
+    try:
+        # DDL format, "fieldname datatype, fieldname datatype".
+        return from_ddl_schema(s)
+    except Exception as e:
+        try:
+            # For backwards compatibility, "integer", "struct<fieldname: datatype>" and etc.
+            return from_ddl_datatype(s)
+        except:
+            try:
+                # For backwards compatibility, "fieldname: datatype, fieldname: datatype" case.
```
Comment (cloud-fan) [inline python/pyspark/sql/types.py]: won't `fieldname: datatype, fieldname: datatype` be parsed as DDL schema?

File: python/pyspark/sql/functions.py [hunk 4]
```diff
@@ -2026,16 +2026,26 @@ def __init__(self, func, returnType, name=None):
...
+        self._returnType = returnType
         # Stores UserDefinedPythonFunctions jobj, once initialized
+        self._returnType_placeholder = None
         self._judf_placeholder = None
         self._name = name or (
             func.__name__ if hasattr(func, '__name__')
             else func.__class__.__name__)
 
     @property
+    def returnType(self):
```
Comment (holdenk) [inline python/pyspark/sql/functions.py]: We have pretty similar logic bellow, would it make sense to think about if there is a nicer more general way to handle these delayed iniatilization classes?

**Review by holdenk — COMMENTED** [review]
Thanks for working on this, unifiying the parsing logic to be Scala side seems like a good idea.

File: python/pyspark/sql/types.py (hunk 3 above)
Comment (HyukjinKwon) [inline python/pyspark/sql/types.py]: I tested few cases but it looks not:

```scala
scala> StructType.fromDDL("a struct<a: INT, b: STRING>")
res5: org.apache.spark.sql.types.StructType = StructType(StructField(a,StructType(StructField(a,IntegerType,true), StructField(b,StringType,true)),true))

scala> StructType.fromDDL("a INT, b STRING")
res6: org.apache.spark.sql.types.StructType = StructType(StructField(a,IntegerType,true), StructField(b,StringType,true))

scala> StructType.fromDDL("a: INT, b: STRING")
org.apache.spark.sql.catalyst.parser.ParseException:
extraneous input ':' expecting ...
```

File: python/pyspark/sql/functions.py (hunk 4 above)
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: hmm.. I tried several ways I could think at my best but I could not figure out ...

**Comment by cloud-fan** [comment]
LGTM, merging to master!

**Pull request closed without merging**